### ISSUE TITLE
QXcbConnection: Could not connect to display

### ISSUE THREAD

**Issue by alfredopalhares** [wkhtmltopdf/wkhtmltopdf#2037]
Hello,

On a headless server this happens:

```
# /usr/bin/wkhtmltopdf --lowquality '/tmp/knp_snappy545b9647bf8b52.57407213.html' '/tmp/knp_snappy545b9647bf90a2.97997365.pdf'
QXcbConnection: Could not connect to display 
Aborted
```

Version

```
# dpkg -s wkhtmltopdf
Package: wkhtmltopdf
Status: install ok installed
Priority: extra
Section: utils
Installed-Size: 979
Maintainer: Emmanuel Bouthenot <kolter@debian.org>
Architecture: amd64
Version: 0.12.1-2
Depends: libc6 (>= 2.14), libgcc1 (>= 1:4.1.1), libgl1-mesa-glx | libgl1, libqt5core5a (>= 5.0.2), libqt5gui5 (>= 5.2.0), libqt5network5 (>= 5.0.2), libqt5printsupport5 (>= 5.2.0), libqt5svg5 (>= 5.0.2), libqt5webkit5 (>= 5.2.0), libqt5widgets5 (>= 5.0.2), libqt5xmlpatterns5 (>= 5.0.2), libstdc++6 (>= 4.1.1)
Recommends: xserver | xvfb
Description: Command line utilities to convert html to pdf or image using WebKit
 wkhtmltopdf is a command line program which permits one to create a
 pdf or an image from an url, a local html file or stdin. It produces a pdf or
 an image like rendred with the WebKit engine.
 .
 This program requires an X11 server to run.
Homepage: http://wkhtmltopdf.org/
```

```
# dpkg --get-selections | grep xvfb
xvfb                                            install
```

Its a Debian machine 3.16.3-2 kernel

Any Ideas ?

**Comment by liotier** [comment]
[The Debian package now recommends xserver](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=580226#34). The recommended workaround in Debian is xvfb.

**Comment by jspizziri** [comment]
@ashkulz, would it be possible to document *why* the debian repos don't support running headless?

**Comment by AlbertoGP** [comment]
The reason from Debian (same in Ubuntu apparently) is mentioned in the bug report:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=580226#10
> Actually, wkhtmltopdf needs a patched QT to work without X. This patch is provided by the upstream developer but it is not suitable for Debian.

**Comment by dustinalan** [comment]
Works great under our Ubuntu 16.04 servers.

See post from @masterkorp above:

> `$ xvfb-run -- /usr/bin/wkhtmltopdf`

**Comment by happyhq** [comment]
thank you!

**Comment by mckaygerhard** [comment]
what happened for this bug today? with qt5?

**Comment by gregogalante** [comment]
The bug is already present installing wkhtmltopdf from apt-get on Ubuntu 16.04

**Comment by mckaygerhard** [comment]
winbuntu? its still need a patched qt to compile?

**Comment by joshuapinter** [comment]
@gregogalante 

### Steps for install 0.12.3 on Ubuntu 16.04.

Because I wasted way too much time tracking this down, here is for others stumbling on this: 

Just to help others, here is what my installation steps were installing 0.12.3 on Ubuntu 16.04: 

```bash
cd ~
wget https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/0.12.3/wkhtmltox-0.12.3_linux-generic-amd64.tar.xz
tar vxf wkhtmltox-0.12.3_linux-generic-amd64.tar.xz 
cp wkhtmltox/bin/wk* /usr/local/bin/
```

And you can confirm with:

```bash
$ wkhtmltopdf --version
wkhtmltopdf 0.12.3 (with patched qt)
```

Btw, I chose 0.12.3 instead of 0.12.4 because of a strange page sizing issue, discussed here: 
https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3226

**Comment by lobermann** [comment]
Above did not work for me on ubuntu 18. I solved it the following way
https://gist.github.com/lobermann/ca0e7bb2558b3b08923c6ae2c37a26ce

**Comment by mckaygerhard** [comment]
this software now its a crap.. breack upstream normal sources and added several changes not officially merged in upstream.. 

today there's qt5 and this cannot run in healess without several changes in the sources those changes not present at upstream qt sources

**Comment by lobermann** [comment]
@mckaygerhard so what are the alternatives? I have tried multiple solutions but this ended up to, overall, produce the most reliable PDF outputs, especially when generating multipage PDFs with tables and page breaks.

**Comment by Toub** [comment]
@joshuapinter solution also work for debian jessie in docker:

```bash
wget --quiet https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/0.12.3/wkhtmltox-0.12.3_linux-generic-amd64.tar.xz && \
    tar vxf wkhtmltox-0.12.3_linux-generic-amd64.tar.xz && \
    cp wkhtmltox/bin/wk* /usr/local/bin/ && \
    rm -rf wkhtmltox
```
Thanks!

**Comment by altjx** [comment]
Has anyone been able to figure out why running `wkhtmltopdf` with `xvfb-run` ignores the font css within the HTML file generated by `wkhtmltopdf`? For example, running `wkhtmltopdf` works just fine and the font css works, but running `xvfb-run` with `wkhtmltopdf` completely ignores the font and the PDF comes out in some Courier font or something. That's the only problem I have now -- spent half of yesterday troubleshooting.

For me, I can't generate charts without running `xvfb-run` with `wkhtmltopdf` from within my docker image so if I could fix the font issue, I'd be in good shape.

**Comment by craigds** [comment]
@altjx did you figure that out? I have the same issue. Fonts get ignored when running via `xvfb-run`. I have installed the fonts but they don't get used.

**Comment by craigds** [comment]
Using the `wkhtmltox` debs from https://github.com/wkhtmltopdf/wkhtmltopdf/releases/ solves the font issue and doesn't require running it in xvfb-run either, so we'll roll with that.

**Comment by belarba** [comment]
> Using the `wkhtmltox` debs from https://github.com/wkhtmltopdf/wkhtmltopdf/releases/ solves the font issue and doesn't require running it in xvfb-run either, so we'll roll with that.

how I use this `wkhtmltox` on rails? The wkhtmltopdf works on AWS with the xvfb-run, but when I put `use_xvfb: true` on wicjed_pdf.rb, it doesnt work.

**Comment by joshuapinter** [comment]
### Installing 0.12.6 on Ubuntu 18.

We've since upgraded to 0.12.6, which fixed the issues found on 0.12.4 and 0.12.5.

Here are our updated installation instructions: 

1. Check what version, if any, you currently have: 

    ```bash
    $ wkhtmltopdf --version
    wkhtmltopdf 0.12.3 (with patched qt)
    ```

2. Remove existing version by finding all files with "wkhtml" and removing them: 

    ```bash
    sudo find / -name *wkhtml*
    ```

    _Remove "wkhtmltopdf" and "wkhtmltox" related files._

3. Download the `.deb` package for version 0.12.6: 

    ```bash
    wget https://github.com/wkhtmltopdf/packaging/releases/download/0.12.6-1/wkhtmltox_0.12.6-1.bionic_amd64.deb
    ```

4. Install the `.deb` package: 

    ```bash
    sudo dpkg -i wkhtmltox_0.12.6-1.bionic_amd64.deb
    ```

    _If you run into an error with missing dependencies, install the dependencies with `sudo apt install -f`._

5. Verify the correct version has been installed and is accessible: 

    ```bash
    $ wkhtmltopdf --version
    wkhtmltopdf 0.12.6 (with patched qt)
    ```

6. Party! 🎉

**Comment by belarba** [comment]
Well, I just found my AWS is using Debian 8 and the newer version of wkhtmltopdf is on Debian 10. So I need to figure out how to make it works on 0.12.1 :/

**Comment by yuki-kimoto** [comment]
> ### Installing 0.12.6 on Ubuntu 18.
> We've since upgraded to 0.12.6, which fixed the issues found on 0.12.4 and 0.12.5.
> 
> Here are our updated installation instructions:
> 
> 1. Check what version, if any, you currently have:
>    ```shell
>    $ wkhtmltopdf --version
>    wkhtmltopdf 0.12.3 (with patched qt)
>    ```
> 2. Remove existing version by finding all files with "wkhtml" and removing them:
>    ```shell
>    sudo find / -name *wkhtml*
>    ```
>    
>    
>    _Remove "wkhtmltopdf" and "wkhtmltox" related files._
> 3. Download the `.deb` package for version 0.12.6:
>    ```shell
>    wget https://github.com/wkhtmltopdf/packaging/releases/download/0.12.6-1/wkhtmltox_0.12.6-1.bionic_amd64.deb
>    ```
> 4. Install the `.deb` package:
>    ```shell
>    sudo dpkg -i wkhtmltox_0.12.6-1.bionic_amd64.deb
>    ```
>    
>    
>    _If you run into an error with missing dependencies, install the dependencies with `sudo apt install -f`._
> 5. Verify the correct version has been installed and is accessible:
>    ```shell
>    $ wkhtmltopdf --version
>    wkhtmltopdf 0.12.6 (with patched qt)
>    ```
> 6. Party! 🎉

I try it, but I can't install wkhtmltopdf.

```
myapp@tk2-250-34842:~$ sudo dpkg -i wkhtmltox_0.12.6-1.bionic_amd64.deb
(Reading database ... 133265 files and directories currently installed.)
Preparing to unpack wkhtmltox_0.12.6-1.bionic_amd64.deb ...
Unpacking wkhtmltox (1:0.12.6-1.bionic) over (1:0.12.6-1.bionic) ...
Setting up wkhtmltox (1:0.12.6-1.bionic) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
Processing triggers for libc-bin (2.27-3ubuntu1) ...
```

```
myapp@tk2-250-34842:~$ wkhtmltopdf --version
-bash: /usr/bin/wkhtmltopdf: No such file or directory
```

Environment.

```
myapp@tk2-250-34842:~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.2 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

**Comment by MittalPatel-BTC** [comment]
> @gregogalante
> 
> ### Steps for installing 0.12.3 on Ubuntu 16.04.
> Because I wasted way too much time tracking this down, here are my installation steps for installing 0.12.3 on Ubuntu 16.04:
> 
> ```shell
> cd ~
> wget https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/0.12.3/wkhtmltox-0.12.3_linux-generic-amd64.tar.xz
> tar vxf wkhtmltox-0.12.3_linux-generic-amd64.tar.xz 
> cp wkhtmltox/bin/wk* /usr/local/bin/
> ```
> 
> And you can confirm with:
> 
> ```shell
> $ wkhtmltopdf --version
> wkhtmltopdf 0.12.3 (with patched qt)
> ```
> 
> Btw, I chose 0.12.3 instead of 0.12.4 because of a strange page sizing issue, discussed here:
> #3226

@joshuapinter: Thanks it's really helpful to me. This fixed my issue! :+1:
